crates/evm/src/evm/db.rs

```diff
@@ -92,13 +110,31 @@ impl<'a, C: sov_modules_api::Context> Database for EvmDb<'a, C> {
 
     fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
         // TODO move to new_raw_with_hash for better performance
-        let code = self
-            .code
-            .get(&code_hash, &mut self.working_set.offchain_state());
+
+        // If CANCUN or later forks are activated, try to fetch code from offchain storage
+        // first. This is to prevent slower lookups in `code`.
+        if self.current_spec.is_enabled_in(CANCUN) {
+            if let Some(code) = self
+                .offchain_code
+                .get(&code_hash, &mut self.working_set.offchain_state())
+            {
+                self.check_against_code_hash(&code, &code_hash)?;
+                return Ok(code);
+            }
+        }
+        let code = self.code.get(&code_hash, &mut self.working_set);
         if let Some(code) = code {
-            if code_hash != keccak256(code.original_bytes()) {
-                return Err(DBError::CodeHashMismatch);
+            // Gradually migrate contract codes into the offchain code state map.
+            if self
+                .offchain_code
```

Comment:
the first part of the if is unnecessary, since we check it in the if before actually